File: benchmark/db/entry.c
/*
 * JULEA - Flexible storage framework
 * Copyright (C) 2019-2020 Michael Straßberger
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <julea-config.h>

#include <glib.h>

#include <string.h>

#include <julea.h>
#include <julea-db.h>

#include "benchmark.h"

#include "common.c"

static void
_benchmark_db_delete(BenchmarkRun* run, gchar const* namespace, gboolean use_batch, gboolean use_index_all, gboolean use_index_single)
{
	gboolean ret;
	g_autoptr(JBatch) delete_batch = NULL;
	g_autoptr(JBatch) batch = NULL;
	g_autoptr(JSemantics) semantics = NULL;
	g_autoptr(GError) b_s_error = NULL;
	g_autoptr(JDBSchema) b_scheme = NULL;

	semantics = j_benchmark_get_semantics();
	delete_batch = j_batch_new(semantics);
	batch = j_batch_new(semantics);

	b_scheme = _benchmark_db_prepare_scheme(namespace, false, use_index_all, use_index_single, batch, delete_batch);

	g_assert_nonnull(b_scheme);
	g_assert_nonnull(run);

	while (j_benchmark_iterate(run))
	{
		_benchmark_db_insert(NULL, b_scheme, NULL, true, false, false, false);

		j_benchmark_timer_start(run);

		for (gint i = 0; i < ((use_index_all || use_index_single) ? N : (N / N_GET_DIVIDER)); i++)
		{
			g_autoptr(JDBEntry) entry = j_db_entry_new(b_scheme, &b_s_error);
			g_autofree gchar* string = _benchmark_db_get_identifier(i);
			g_autoptr(JDBSelector) selector = j_db_selector_new(b_scheme, J_DB_SELECTOR_MODE_AND, &b_s_error);

			ret = j_db_selector_add_field(selector, "string", J_DB_SELECTOR_OPERATOR_EQ, string, 0, &b_s_error);
			g_assert_true(ret);
			g_assert_null(b_s_error);

			ret = j_db_entry_delete(entry, selector, batch, &b_s_error);
			g_assert_true(ret);
			g_assert_null(b_s_error);

			if (!use_batch)
			{
				ret = j_batch_execute(batch);
				g_assert_true(ret);
			}
		}

		if (use_batch)
		{
			ret = j_batch_execute(batch);
			g_assert_true(ret);
		}

		j_benchmark_timer_stop(run);
	}

	ret = j_batch_execute(delete_batch);
	g_assert_true(ret);

	run->operations = ((use_index_all || use_index_single) ? N : (N / N_GET_DIVIDER));
}

static void
_benchmark_db_update(BenchmarkRun* run, gchar const* namespace, gboolean use_batch, gboolean use_index_all, gboolean use_index_single)
{
	gboolean ret;
	g_autoptr(JBatch) delete_batch = NULL;
	g_autoptr(JBatch) batch = NULL;
	g_autoptr(JSemantics) semantics = NULL;
	g_autoptr(GError) b_s_error = NULL;
	g_autoptr(JDBSchema) b_scheme = NULL;

	semantics = j_benchmark_get_semantics();
	delete_batch = j_batch_new(semantics);
	batch = j_batch_new(semantics);

	b_scheme = _benchmark_db_prepare_scheme(namespace, false, use_index_all, use_index_single, batch, delete_batch);

	g_assert_nonnull(b_scheme);
	g_assert_nonnull(run);

	_benchmark_db_insert(NULL, b_scheme, NULL, true, false, false, false);

	j_benchmark_timer_start(run);

	while (j_benchmark_iterate(run))
	{
		for (gint i = 0; i < ((use_index_all || use_index_single) ? N : (N / N_GET_DIVIDER)); i++)
		{
			gint64 i_signed = (((i + N_PRIME) * SIGNED_FACTOR) % CLASS_MODULUS) - CLASS_LIMIT;
			g_autoptr(JDBSelector) selector = j_db_selector_new(b_scheme, J_DB_SELECTOR_MODE_AND, &b_s_error);
			g_autoptr(JDBEntry) entry = j_db_entry_new(b_scheme, &b_s_error);
			g_autofree gchar* string = _benchmark_db_get_identifier(i);

			g_assert_null(b_s_error);

			ret = j_db_entry_set_field(entry, "sint", &i_signed, 0, &b_s_error);
			g_assert_true(ret);
			g_assert_null(b_s_error);

			ret = j_db_selector_add_field(selector, "string", J_DB_SELECTOR_OPERATOR_EQ, string, 0, &b_s_error);
			g_assert_true(ret);
			g_assert_null(b_s_error);

			ret = j_db_entry_update(entry, selector, batch, &b_s_error);
			g_assert_true(ret);
			g_assert_null(b_s_error);

			if (!use_batch)
			{
				ret = j_batch_execute(batch);
				g_assert_true(ret);
			}
		}

		if (use_batch)
		{
			ret = j_batch_execute(batch);
			g_assert_true(ret);
		}
	}

	j_benchmark_timer_stop(run);

	ret = j_batch_execute(delete_batch);
	g_assert_true(ret);

	run->operations = ((use_index_all || use_index_single) ? N : (N / N_GET_DIVIDER));
}

static void
benchmark_db_insert(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert", false, false, false, true);
}

static void
benchmark_db_insert_batch(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_batch", true, false, false, true);
}

static void
benchmark_db_insert_index_single(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_index_single", false, false, true, true);
}

static void
benchmark_db_insert_batch_index_single(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_batch_index_single", true, false, true, true);
}

static void
benchmark_db_insert_index_all(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_index_all", false, true, false, true);
}

static void
benchmark_db_insert_batch_index_all(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_batch_index_all", true, true, false, true);
}

static void
benchmark_db_insert_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_index_mixed", false, true, true, true);
}

static void
benchmark_db_insert_batch_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_insert(run, NULL, "benchmark_insert_batch_index_mixed", true, true, true, true);
}

static void
benchmark_db_delete(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete", false, false, false);
}

static void
benchmark_db_delete_batch(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_batch", true, false, false);
}

static void
benchmark_db_delete_index_single(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_index_single", false, false, true);
}

static void
benchmark_db_delete_batch_index_single(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_batch_index_single", true, false, true);
}

static void
benchmark_db_delete_index_all(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_index_all", false, true, false);
}

static void
benchmark_db_delete_batch_index_all(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_batch_index_all", true, true, false);
}

static void
benchmark_db_delete_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_index_mixed", false, true, true);
}

static void
benchmark_db_delete_batch_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_delete(run, "benchmark_delete_batch_index_mixed", true, true, true);
}

static void
benchmark_db_update(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update", false, false, false);
}

static void
benchmark_db_update_batch(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_batch", true, false, false);
}

static void
benchmark_db_update_index_single(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_index_single", false, false, true);
}

static void
benchmark_db_update_batch_index_single(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_batch_index_single", true, false, true);
}

static void
benchmark_db_update_index_all(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_index_all", false, true, false);
}

static void
benchmark_db_update_batch_index_all(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_batch_index_all", true, true, false);
}

static void
benchmark_db_update_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_index_mixed", false, true, true);
}

static void
benchmark_db_update_batch_index_mixed(BenchmarkRun* run)
{
	_benchmark_db_update(run, "benchmark_update_batch_index_mixed", true, true, true);
}

void
benchmark_db_entry(void)
{
	j_benchmark_add("/db/entry/insert", benchmark_db_insert);
	j_benchmark_add("/db/entry/insert-batch", benchmark_db_insert_batch);
	j_benchmark_add("/db/entry/insert-index-single", benchmark_db_insert_index_single);
	j_benchmark_add("/db/entry/insert-batch-index-single", benchmark_db_insert_batch_index_single);
	j_benchmark_add("/db/entry/insert-index-all", benchmark_db_insert_index_all);
	j_benchmark_add("/db/entry/insert-batch-index-all", benchmark_db_insert_batch_index_all);
	j_benchmark_add("/db/entry/insert-index-mixed", benchmark_db_insert_index_mixed);
	j_benchmark_add("/db/entry/insert-batch-index-mixed", benchmark_db_insert_batch_index_mixed);
	j_benchmark_add("/db/entry/delete", benchmark_db_delete);
	j_benchmark_add("/db/entry/delete-batch", benchmark_db_delete_batch);
	j_benchmark_add("/db/entry/delete-index-single", benchmark_db_delete_index_single);
	j_benchmark_add("/db/entry/delete-batch-index-single", benchmark_db_delete_batch_index_single);
	j_benchmark_add("/db/entry/delete-index-all", benchmark_db_delete_index_all);
	j_benchmark_add("/db/entry/delete-batch-index-all", benchmark_db_delete_batch_index_all);
	j_benchmark_add("/db/entry/delete-index-mixed", benchmark_db_delete_index_mixed);
	j_benchmark_add("/db/entry/delete-batch-index-mixed", benchmark_db_delete_batch_index_mixed);
	j_benchmark_add("/db/entry/update", benchmark_db_update);
	j_benchmark_add("/db/entry/update-batch", benchmark_db_update_batch);
	j_benchmark_add("/db/entry/update-index-single", benchmark_db_update_index_single);
	j_benchmark_add("/db/entry/update-batch-index-single", benchmark_db_update_batch_index_single);
	j_benchmark_add("/db/entry/update-index-all", benchmark_db_update_index_all);
	j_benchmark_add("/db/entry/update-batch-index-all", benchmark_db_update_batch_index_all);
	j_benchmark_add("/db/entry/update-index-mixed", benchmark_db_update_index_mixed);
	j_benchmark_add("/db/entry/update-batch-index-mixed", benchmark_db_update_batch_index_mixed);
}

File: benchmark/python/db/entry.py
from benchmarkrun import BenchmarkRun, append_to_benchmark_list_and_run
from julea import lib, encode, ffi
from db.common import _benchmark_db_insert, _benchmark_db_prepare_scheme, _benchmark_db_get_identifier, N, N_GET_DIVIDER, N_PRIME, SIGNED_FACTOR, CLASS_MODULUS, CLASS_LIMIT

def benchmark_db_entry(benchmarkrun_list, iterations, machine_readable):
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert", iterations, machine_readable), benchmark_db_insert)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-batch", iterations, machine_readable), benchmark_db_insert_batch)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-index-single", iterations, machine_readable), benchmark_db_insert_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-batch-index-single", iterations, machine_readable), benchmark_db_insert_batch_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-index-all", iterations, machine_readable), benchmark_db_insert_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-batch-index-all", iterations, machine_readable), benchmark_db_insert_batch_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-index-mixed", iterations, machine_readable), benchmark_db_insert_index_mixed)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/insert-batch-index-mixed", iterations, machine_readable), benchmark_db_insert_batch_index_mixed)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete", iterations, machine_readable), benchmark_db_delete)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-batch", iterations, machine_readable), benchmark_db_delete_batch)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-index-single", iterations, machine_readable), benchmark_db_delete_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-batch-index-single", iterations, machine_readable), benchmark_db_delete_batch_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-index-all", iterations, machine_readable), benchmark_db_delete_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-batch-index-all", iterations, machine_readable), benchmark_db_delete_batch_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-index-mixed", iterations, machine_readable), benchmark_db_delete_index_mixed)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/delete-batch-index-mixed", iterations, machine_readable), benchmark_db_delete_batch_index_mixed)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update", iterations, machine_readable), benchmark_db_update)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-batch", iterations, machine_readable), benchmark_db_update_batch)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-index-single", iterations, machine_readable), benchmark_db_update_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-batch-index-single", iterations, machine_readable), benchmark_db_update_batch_index_single)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-index-all", iterations, machine_readable), benchmark_db_update_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-batch-index-all", iterations, machine_readable), benchmark_db_update_batch_index_all)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-index-mixed", iterations, machine_readable), benchmark_db_update_index_mixed)
    append_to_benchmark_list_and_run(benchmarkrun_list, BenchmarkRun("/db/entry/update-batch-index-mixed", iterations, machine_readable), benchmark_db_update_batch_index_mixed)

def benchmark_db_insert(run):
    _benchmark_db_insert(run, None, "benchmark_insert", False, False, False,
                         True)

def benchmark_db_insert_batch(run):
    _benchmark_db_insert(run, None, "benchmark_insert_batch", True, False,
                         False, True)

def benchmark_db_insert_index_single(run):
    _benchmark_db_insert(run, None, "benchmark_insert_index_single", False,
                         False, True, True)

def benchmark_db_insert_batch_index_single(run):
    _benchmark_db_insert(run, None, "benchmark_insert_batch_index_single", True,
                         False, True, True)

def benchmark_db_insert_index_all(run):
    _benchmark_db_insert(run, None, "benchmark_insert_index_all", False, True,
                         False, True)

def benchmark_db_insert_batch_index_all(run):
    _benchmark_db_insert(run, None, "benchmark_insert_batch_index_all", True,
                         True, False, True)

def benchmark_db_insert_index_mixed(run):
    _benchmark_db_insert(run, None, "benchmakr_insert_index_mixed", False, True,
                         True, True)

def benchmark_db_insert_batch_index_mixed(run):
    _benchmark_db_insert(run, None, "benchmakr_insert_batch_index_mixed", True,
                         True, True, True)

def benchmark_db_delete(run):
    _benchmark_db_delete(run, "benchmark_delete", False, False, False)

def benchmark_db_delete_batch(run):
    _benchmark_db_delete(run, "benchmark_delete_batch", True, False, False)

def benchmark_db_delete_index_single(run):
    _benchmark_db_delete(run, "benchmark_delete_index_single", False, False,
                         True)

def benchmark_db_delete_batch_index_single(run):
    _benchmark_db_delete(run, "benchmark_delete_batch_index_single", True,
                         False, True)

def benchmark_db_delete_index_all(run):
    _benchmark_db_delete(run, "benchmark_delete_index_all", False, True, False)

def benchmark_db_delete_batch_index_all(run):
    _benchmark_db_delete(run, "benchmark_delete_batch_index_all", True, True,
                         False)

def benchmark_db_delete_index_mixed(run):
    _benchmark_db_delete(run, "benchmark_delete_index_mixed", False, True, True)

def benchmark_db_delete_batch_index_mixed(run):
    _benchmark_db_delete(run, "benchmark_delete_batch_index_mixed", True, True,
                         True)

def _benchmark_db_delete(run, namespace, use_batch, use_index_all,
                         use_index_single):
    namespace_encoded = encode(namespace)
    b_s_error_ptr = ffi.new("GError*")
    b_s_error_ptr = ffi.NULL
    batch = lib.j_batch_new_for_template(lib.J_SEMANTICS_TEMPLATE_DEFAULT)
    delete_batch = lib.j_batch_new_for_template(lib.J_SEMANTICS_TEMPLATE_DEFAULT)
    b_scheme = _benchmark_db_prepare_scheme(namespace_encoded, False,
                                            use_index_all, use_index_single,
                                            batch, delete_batch)
    assert b_scheme != ffi.NULL
    assert run != None
    _benchmark_db_insert(None, b_scheme, "\0", True, False, False, False)
    run.start_timer()
    iterations = N if use_index_all or use_index_single else int(N / N_GET_DIVIDER)
    for i in range(iterations):
        entry = lib.j_db_entry_new(b_scheme, b_s_error_ptr)
        string = encode(_benchmark_db_get_identifier(i))
        string_name = encode("string")
        selector = lib.j_db_selector_new(b_scheme, lib.J_DB_SELECTOR_MODE_AND,
                                         b_s_error_ptr)
        assert lib.j_db_selector_add_field(selector, string_name,
                                           lib.J_DB_SELECTOR_OPERATOR_EQ,
                                           string, 0, b_s_error_ptr)
        assert b_s_error_ptr == ffi.NULL
        assert lib.j_db_entry_delete(entry, selector, batch, b_s_error_ptr)
        assert b_s_error_ptr == ffi.NULL
        if not use_batch:
            assert lib.j_batch_execute(batch)
        lib.j_db_entry_unref(entry)
        lib.j_db_selector_unref(selector)
    if use_batch:
            assert lib.j_batch_execute(batch)
    run.stop_timer()
    assert lib.j_batch_execute(delete_batch)
    run.operations = iterations
    lib.j_batch_unref(batch)
    lib.j_batch_unref(delete_batch)
    lib.j_db_schema_unref(b_scheme)

def benchmark_db_update(run):
    _benchmark_db_update(run, "benchmark_update", False, False, False)

def benchmark_db_update_batch(run):
    _benchmark_db_update(run, "benchmark_update_batch", True, False, False)

def benchmark_db_update_index_single(run):
    _benchmark_db_update(run, "benchmark_update_index_single", False, False,
                         True)

def benchmark_db_update_batch_index_single(run):
    _benchmark_db_update(run, "benchmark_update_batch_index_single", True,
                         False, True)

def benchmark_db_update_index_all(run):
    _benchmark_db_update(run, "benchmark_update_index_all", False, True, False)

def benchmark_db_update_batch_index_all(run):
    _benchmark_db_update(run, "benchmark_update_batch_index_all", True, True,
                         False)

def benchmark_db_update_index_mixed(run):
    _benchmark_db_update(run, "benchmark_update_index_mixed", False, True, True)

def benchmark_db_update_batch_index_mixed(run):
    _benchmark_db_update(run, "benchmark_update_batch_index_mixed", True, True,
                         True)

def _benchmark_db_update(run, namespace, use_batch, use_index_all,
                         use_index_single):
    namespace_encoded = encode(namespace)
    b_s_error_ptr = ffi.new("GError*")
    b_s_error_ptr = ffi.NULL
    batch = lib.j_batch_new_for_template(lib.J_SEMANTICS_TEMPLATE_DEFAULT)
    delete_batch = lib.j_batch_new_for_template(lib.J_SEMANTICS_TEMPLATE_DEFAULT)
    b_scheme = _benchmark_db_prepare_scheme(namespace_encoded, False,
                                            use_index_all, use_index_single,
                                            batch, delete_batch)
    assert b_scheme != ffi.NULL
    assert run != None
    _benchmark_db_insert(None, b_scheme, "\0", True, False, False, False)
    run.start_timer()
    iterations = N if use_index_all or use_index_single else int(N / N_GET_DIVIDER)
    for i in range(iterations):
        sint_name = encode("sint")
        i_signed_ptr = ffi.new("long*")
        i_signed_ptr[0] = (((i + N_PRIME) * SIGNED_FACTOR) & CLASS_MODULUS) - CLASS_LIMIT
        selector = lib.j_db_selector_new(b_scheme, lib.J_DB_SELECTOR_MODE_AND,
                                         b_s_error_ptr)
        entry = lib.j_db_entry_new(b_scheme, b_s_error_ptr)
        string_name = encode("string")
        string = encode(_benchmark_db_get_identifier(i))
        assert b_s_error_ptr == ffi.NULL
        assert lib.j_db_entry_set_field(encode, sint_name, i_signed_ptr, 0,
                                        b_s_error_ptr)
        assert b_s_error_ptr == ffi.NULL
        assert lib.j_db_selector_add_field(selector, string_name,
                                           lib.J_DB_SELECTOR_OPERATOR_EQ,
                                           string, 0, b_s_error_ptr)
        assert b_s_error_ptr == ffi.NULL
        assert lib.j_db_entry_update(entry, selector, batch, b_s_error_ptr)
        assert b_s_error_ptr == ffi.NULL
        if not use_batch:
            assert lib.j_batch_execute(batch)
        lib.j_db_selector_unref(selector)
        lib.j_db_entry_unref(entry)
    if use_batch:
        assert lib.j_batch_execute(batch)
    run.stop_timer()
    assert lib.j_batch_execute(delete_batch)
    run.operations = iterations
    lib.j_batch_unref(batch)
    lib.j_batch_unref(delete_batch)
    lib.j_db_schema_unref(b_scheme)
